stream-chat-android-offline/src/test/java/io/getstream/chat/android/livedata/BaseDomainTest2.kt

```diff
@@ -0,0 +1,258 @@
+ * - Use Executors.newSingleThreadExecutor to prevent room transactions from causing deadlocks
+ * execute instead of running on a different thread
+ *
+ * Handling Room and Coroutines is unfortunately quite tricky.
+ * * https://www.youtube.com/watch?v=KMb0Fs8rCRs&feature=youtu.be
+ * * https://medium.com/@eyalg/testing-androidx-room-kotlin-coroutines-2d1faa3e674f
+ * * https://medium.com/androiddevelopers/threading-models-in-coroutines-and-android-sqlite-api-6cab11f7eb90
+ * * https://kotlin.github.io/kotlinx.coroutines/kotlinx-coroutines-test/
+ * * https://codelabs.developers.google.com/codelabs/kotlin-coroutines#9
+ * * https://craigrussell.io/2019/11/unit-testing-coroutine-suspend-functions-using-testcoroutinedispatcher/
+ *
+ */
+internal open class BaseDomainTest2 {
+
+    /** a realistic set of chat data, please only add to this, don't update */
+    var data = TestDataHelper()
+
+    /** the chat domain impl */
+    lateinit var chatDomainImpl: ChatDomainImpl
+    /** the chat domain interface */
+    lateinit var chatDomain: ChatDomain
+
+    /** the mock for the chat client */
+    lateinit var clientMock: ChatClient
+
+    /** a channel controller for data.channel1 */
+    lateinit var channelControllerImpl: ChannelControllerImpl
+
+    /** a queryControllerImpl for the query */
+    lateinit var queryControllerImpl: QueryChannelsControllerImpl
+    /** the query used for the default queryController */
+    lateinit var query: QueryChannelsSpec
+
+    /** a mock for the channel client */
+    lateinit var channelClientMock: ChannelClient
+
+    private lateinit var db: ChatDatabase
+
+    // the code below is used to ensure that coroutines execute the right way during tests
+    private val testIODispatcher = TestCoroutineDispatcher()
+    private val testIOExecutor = testIODispatcher.asExecutor()
+    internal val testIOScope = TestCoroutineScope(testIODispatcher)
+
+    private val testMainDispatcher = TestCoroutineDispatcher()
+    private val testMainExecutor = testMainDispatcher.asExecutor()
+    internal val testMainScope = TestCoroutineScope(testMainDispatcher)
+
+    val singleThreadExecutor = Executors.newSingleThreadExecutor()
+
+    /**
+     * checks if a response is succesful and raises a clear error message if it's not
+     */
+    fun assertSuccess(result: Result<*>) {
+        if (result.isError) {
+            Truth.assertWithMessage(result.error().toString()).that(result.isError).isFalse()
+        }
+    }
+
+    /**
+     * checks if a response failed and raises a clear error message if it succeeded
+     */
+    fun assertFailure(result: Result<*>) {
+        if (!result.isError) {
+            Truth.assertWithMessage(result.data().toString()).that(result.isError).isTrue()
+        }
+    }
+
+    @get:Rule
+    val rule = InstantTaskExecutorRule()
+
+    @Before
+    open fun setup() {
+        clientMock = createClientMock()
```

Comment:
In test, it's also better to set the main Dispatcher. 
``` kotlin
...
  val testDispatcher = TestCoroutineDispatcher()
...
@Before
fun setup() {
  Dispatchers.setMain(testCoroutineDispatcher)
  ...
}

@After
fun tearDown() {
  Dispatchers.resetMain()
  ...
}

```

Reply:
yup, added it